CIP-0032/README.md

```diff
@@ -0,0 +1,129 @@
+
+### UTXO set size
+
+This proposal gives users a way to put much larger amounts of data into the UTXO set. 
+Won’t this lead to much worse UTXO set bloat?
+
+The answer is that we already have a mechanism to discourage this, namely the minimum UTXO value. 
+If inline datums turns out to drive significantly increased space usage, then we may need to increase `coinsPerUTxOWord` in order to keep the UTXO size down.
+That will be costly and inconvenient for users, but will still allow them to use inline datums where they are most useful and the cost is bearable.
+Furthermore, we hope that we will in fact be able to _reduce_ `coinsPerUTxOWord` when the upcoming work on moving the UTXO mostly to on-disk storage is complete.
+
+Another guard rail would be to enforce upper limits on the size of inline datums.
+At the extreme, we could bound them to the size of a hash, which would guarantee no more space usage than today.
+However, this is much worse for users, since it introduces a sharp discontinuity where an inline datum is entirely acceptable, until it crosses the size threshold at which point it is unacceptable, and there is no way to avoid this.
+Generally we prefer to avoid such discontinuities in favour of gradually increasing costs.
+
+In practice, what is implemented here may depend on whether the UTXO-on-disk work is completed by the time that this proposal is implemented.
+
+### Other modes of specifying datums
+
+We could deprecate the other methods of specifying datums (datum hashes or datum hashes+optional datums).
+However, the other approaches also have some advantages.
+
+- Transmission costs: creator pays versus consumer pays
+    - Inline datums: creator pays
+    - Datum hashes: consumer pays
+    - Datum hashes+optional datums: both pay
+- Min UTXO value costs
+    - Inline datums: depends on data size
+    - Datum hashes: fixed cost
+    - Datum hashes+optional datums: fixed cost
+- Privacy
+    - Inline datums: datum is immediately public
+    - Datum hashes: datum is not public until consumed
+    - Datum hashes+optional datums: datum is immediately public, but only to chain-followers
+- Communication of datums
+    - Inline datums: easy on-chain communication
+    - Datum hashes: off-chain communication necessary
+    - Datum hashes+optional datums: complicated on-chain communication
+
+Any one of these factors could be important to particular use cases, so it is good to retain the other options.
+
+### Changing the script context
+
+We don't strictly need to change the script context.
+We could use the fall-back approach that we use for the old interface, namely continuing to pretend that datums are always hashes, and providing the mapping from hashes to datums.
+This would be functionally equivalent.
+
+The only advantage of changing it is that it's more honest to the real representation of transactions, and therefore allows scripts to insist on inline datums if they want.
```

Comment:
For data-availability uses cases being able to enforce the usage of inline datums is crucial so I would definitely make some way to enforce it (be it through its presence in the script context or some other way)

Reply:
This only makes a difference off-chain. IICR validators can already enforce datums for new outputs to be present in a transaction, so the only real difference between inline or not is whether to look it up in the transaction history or the UTXO. Making the inlining status visible to validators would simplify this lookup, at the cost of adding complexity to the on-chain code for everyone.

Reply:
> is whether to look it up in the transaction history or the UTXO

Which is a big deal! So I think Sebastian's point is relevant: people will want to be able to enforce this.

Reply:
@WhatisRT Can you explain how to find datums trustlessly offchain? I can't seem to find any answers online, this issue comes closest

Reply:
Since the hash is published on-chain, there's no requirement for trustlessness here